File: python/question.py
# implementation of the question function in Python
def question(prompt, valid):
    inp = ""
    while True:

        print(prompt)
        if len(valid) != 0:
            inp = input(f'({", ".join(valid)}): ').strip()
        else:
            inp = input(": ").strip()

        if len(valid) == 0 or inp in valid:
            return inp

        print(f'"{inp}" is not a valid answer')


question("foo", ["bar", "baz"])

File: nim/question.nim
import std/[strutils, strbasics, strformat, rdstdin]

# implementation of the question "function" in nim
proc question(prompt: string): string =
  echo prompt
  readLineFromStdin(": ")

proc question[I](prompt: string, valid: array[I, string]): string =
  let joined_prompt = &"""({valid.join(", ")}): """

  while true:
    echo prompt
    var input = readLineFromStdin(joined_prompt)
    input.strip()
    if input in valid: return input
    echo '"', input, '"', " is not a valid answer"

when isMainModule:
  discard question("foo", ["bar", "baz"])
